python/fastfec.py

```diff
@@ -1,59 +1,386 @@
+
+        from fastfec import FastFEC
+
+        with open('12345.fec', 'rb') as f:
+            with FastFEC() as fastfec:
+                for form, line in fastfec.parse(f):
+                    print("GOT", form, line)
+
+    * Parse a filing to output files:
+
+        import os
+        from pathlib import Path
+        from fastfec import FastFEC
+
+        # Custom open method that places files in an output directory,
+        # and creates the directory if it does not exist
+        def open_output_file(filename, *args, **kwargs):
+            filename = os.path.join('custom_python_output', filename)
+            output_file = Path(filename)
+            output_file.parent.mkdir(exist_ok=True, parents=True)
+            return open(filename, *args, **kwargs)
+
+        with open('12345.fec', 'rb') as f:
+            with FastFEC() as fastfec:
+                fastfec.parse_as_files(f, open_output_file)
+"""
+
+import contextlib
+import csv
+import datetime
+import logging
 import os
 import pathlib
-from smart_open import open
+from ctypes import (CDLL, CFUNCTYPE, POINTER, c_char, c_char_p, c_int,
+                    c_size_t, c_void_p, memmove)
+from ctypes.util import find_library
+from dataclasses import dataclass
+from glob import glob
+from queue import Queue
+from threading import Thread
 
+# Directories used for locating the shared library
 SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
+PARENT_DIR = pathlib.Path(SCRIPT_DIR).parent.absolute()
 
+# Buffer constants
 BUFFER_SIZE = 1024 * 1024
-TRANSFER_BUFFER_SIZE = 128 * 1024
+
+# Callback function ctypes
 BUFFER_READ = CFUNCTYPE(c_size_t, POINTER(c_char), c_int, c_void_p)
 CUSTOM_WRITE = CFUNCTYPE(None, c_char_p, c_char_p,
                          c_char_p, c_int)
+CUSTOM_LINE = CFUNCTYPE(None, c_char_p, c_char_p, c_char_p)
+
+
+def make_read_buffer(file_input):
+    """Creates a read buffer callback given an open input stream
+
+    Arguments:
+        file_input -- An input stream for the callback to consume
+
+    Returns:
+        A callback function that C code can wrap to consume the input stream"""
 
-def make_read_buffer(f):
     def read_buffer(buffer, want, _):
-        contents_raw = f.read(want)
+        contents_raw = file_input.read(want)
         received = len(contents_raw)
         contents = c_char_p(contents_raw)
         memmove(buffer, contents, received)
         return received
     return read_buffer
 
 
-class FastFEC:
-    def __init__(self, file_handle, output_directory, use_compression=False, make_dirs=False):
-        self.init_lib()
-        self.file_handle = file_handle
-        self.use_compression=use_compression
-        self.make_dirs = make_dirs
+def find_fastfec_lib():
+    """Scans for the fastfec shared library and returns the path of the found library
+
+    This method tries searching in the parent directory first but has a fallback to
+    the local zig build directory for development work.
+    """
```

Comment:
For context, the Python packaging system will automatically place it in the parent directory, so I _think_ that is the first and only place it will have to be looked pretty much. Technically we could probably have that be the only check, but I wanted it to eagerly try some backups just in case (if you install with brew, it will be found via the system lookup I think, for instance)

Reply:
What about the homebrew install? Or would it just be installed in two places?

Reply:
Or, I guess, I'm not sure what my expectations are here as a user. If I've installed this through homebrew and I install the python package I don't know If am expecting it to install another (perhaps, different) version and use that instead?

Reply:
Will remove looking in system library directory